submariner/OVN-Interconnect.md

```diff
@@ -0,0 +1,239 @@
+  zones. The OVN-Kubernetes services ensure that the necessary routes are added for pod and service reachability across nodes in different
+  zones.
+
+```bash
+    annotations:
+      k8s.ovn.org/ovn-node-transit-switch-port-ips: '["169.254.0.3/16"]'
+      k8s.ovn.org/ovn-zone: global
+    name: cluster1-worker
+
+    annotations:
+      k8s.ovn.org/ovn-node-transit-switch-port-ips: '["169.254.0.5/16"]'
+      k8s.ovn.org/ovn-zone: az2
+    name: cluster2-worker
+```
+
+With the current architecture, Submariner adds routes only in the zone in which it is deployed. For example, if Submariner is deployed in
+zone 1 it programs OVN db in zone 1. So only pods in zone 1 nodes will be able to talk to other clusters. Pods in zone 2 or zone 3 will not
+be able to reach remote clusters connected via Submariner. With these changes we plan to support both the modes and cluster where interconnect
+is not enabled as well.
+
+### Route APIs
+
+As part of this proposal, we are planning to add two new CRDs in Submariner. The plan is to decouple OVN dataplane programming so that this
+can be moved to OVN, when OVN provides a way to add custom routes. We can propose this CRD or consume the CRD that they provide at a later
+point of time.
+
+#### SubmarinerGwRoute
+
+This CR will be created when a remote endpoint is added and there will be one CR per remote cluster. This crd has two fields
+
+* NextHop - Specifies the next hop to reach the remote cluster, in this case it will be the IP of ovn-k8s-mp0
+  interface, the interface used by OVN for host networking.
+* RemoteCIDR - Specifies the list of  remote CIDRs reachable via this cluster.
+
+This CR will be used by the route agent pod running on the non-Gateway nodes to program OVN to send the traffic destined to
+remote clusters via the Submariner tunnel.
+
+``` go
+type SubmarinerGwRoute struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    SubmarinerRoutePolicySpec SubmarinerRoutePolicySpec `json:"submarinerRoutePolicySpec"`
+}
+
+type SubmarinerRoutePolicySpec struct {
+    //Specifies the next hop to reach the remote CIDRs
+    NextHop string `json:"nextHop"`
+
+    //Specifies the remote CIDRs available via the next hop
+    RemoteCidr []string `json:"remoteCidr"`
+}
+```
+
+#### SubmarinerNonGWRoute
+
+This CR will be created when a remote endpoint is created. When more remote endpoints are added the Remote CIDR list will be updated.
+
+* NextHop - Specifies the next hop.
+* RemoteCIDR - Specifies the list of remote CIDRs reachable via this gateway.
+
+This CR will be used by the Routeagent pods running on the non-Gateway nodes to
+
+* In non-g/w node - If the route-agent pod is not in the same zone as Gateway node zone, send the traffic to the g/w node zone.
+* In g/w node - add route to send the traffic from other zones, destined to remote cluster to the submariner tunnel
+
+``` go
+type SubmarinerNonGWRoute struct {
```

Comment:
Why do we need two different CRDs (SubmarinerNonGWRoute, SubmarinerGwRoute)? it seems that the content is identical for both of them.
Maybe we can add type field?

Reply:
I thought having two controller handling it would be easier , but I am open to change 

@sridhargaddam  any thoughts?

Reply:
What will be the value of `NextHop` in this CR and will there not be a difference in this value compared to the other CR (i.e., SubmarinerGwRoute). Can you please clarify?

Reply:
the values will be different, in a SubmarinerNonGWRoute it points to transit switch IP and in SubmarinerGwRoute it points ovn-k8s-mp0 IP. The cluster are CIDRs will be the remote pod and service CIDRs.

The thing common between them is , it has same fields.

Reply:
After the discussion decided to use same CRD.